---
title: "make_fGWAS-input_diagram-1KGenomes.Rmd"
author: "Jason Torres"
date: "February 4, 2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Taking as input credible set files from Anubha's DIAGRAM (1K Genomes-imputed) anaylsis (Accessed Feb.4,2017)


```{r}

"%&%" <- function(a,b) paste0(a,b)
library("data.table")
library("dplyr")
library(GenomicRanges)

serv.dir <- "/Users/jtorres/FUSE/"
gwas.dir <- serv.dir %&% "reference/gwas/diagram_1Kgenomes/"
cred.dir <- gwas.dir %&% "credible_sets/"
cred.files <- list.files(cred.dir)[grepl("_window.txt",list.files(cred.dir))]
names(cred.files) <- 1:length(cred.files)

#bed.dir <- serv.dir %&% "reference/islet/"
save.dir <- serv.dir %&% "projects/t2d-integration/fGWAS_files/diagram_1Kgenomes/"

diag.df <- fread("cat " %&% gwas.dir %&% "DIAGRAM_T2D_1000G.2017.gz")

```


```{r}

build_chunk <- function(refnum){
  fname <- as.character(cred.files[refnum])
  print(fname)
  nam <- gsub("Credset_99_1000G__","",fname)
  nam <- gsub("_201015_LD_window.txt","",nam)
  nam <- gsub("_",":",nam)
  nam <- paste0("chr",nam)
  df <- fread(cred.dir %&% fname)
  
  df <- select(df, one_of("chrpos_b37","n99"))
  names(df) <- c("SNPID","SEGNUMBER")
  df$SNPID <- paste0("chr",df$SNPID)
  CHR <- as.character(sapply(df$SNPID,function(string){
     tmp <- strsplit(string,split=":")[[1]][1]
     return(tmp)
  }))
  POS <- as.integer(sapply(df$SNPID,function(string){
     tmp <- strsplit(string,split=":")[[1]][2]
     return(tmp)
  }))  
  df <- data.frame(df$SNPID,CHR,POS,df$SEGNUMBER,stringsAsFactors=FALSE)
  names(df) <- c("SNPID","CHR","POS","SEGNUMBER")

  df <- arrange(df,POS)
  return(df)
}

build_fgwas_core <- function(){
  out.df <- c() 
  for (i in as.integer(names(cred.files))){
    df <- build_chunk(i)
    out.df <- rbind(out.df,df)
  }
  out.df <- arrange(out.df, SEGNUMBER)
  out.df$CHR  <- paste0("chr",as.character(out.df$CHR))
  return(out.df)
}

core.df <- build_fgwas_core()

```


Functions for getting annotation columns from annotation files 

```{r}


g <- function(df){
   if(length(df)==3){
      gr <- with(df, GRanges(chr, IRanges(start, end)))
   } else if (length(df)==4){
      gr <- with(df, GRanges(chr, IRanges(start, end), id=id))
   } else if (length(df)==5){
      gr <- with(df, GRanges(chr, IRanges(start, end), id=id, score=score))
   } else if (length(df)>=6){
      df$strand <- gsub(pattern="[^+-]+",replacement='*', df$strand)
      gr <- with(df, GRanges(chr, IRanges(start, end), id=id, score=score, strand=strand))
   }
   return(gr)
}

get_overlaps <- function(a,b){
  # a and b are genomic ranges
  val <- as.integer(a %over% b)
  return(val)
}



get_annot <- function(core.df, annot.df, prefix){
  pb <- txtProgressBar(min = 0, max = dim(core.df)[1], initial = 0, style = 3)
  vec <- as.integer(sapply(1:dim(core.df)[1], function(i){
    setTxtProgressBar(pb, i) #print(i) 
    chr <- core.df$CHR[i]
    pos <- core.df$POS[i]
    snp.gr <- GRanges(chr,IRanges(pos, pos))
    annot.gr <- g(annot.df)
    value <- get_overlaps(snp.gr,annot.gr)
    return(value)
  }))
  close(pb)
  saveRDS(vec,file=save.dir%&%prefix%&%".RDS")
  return(vec)
}


```

