---
title: "make_fGWAS-input_diagram-1KGenomes.Rmd"
author: "Jason Torres"
date: "February 4, 2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Taking as input credible set files from Anubha's DIAGRAM (1K Genomes-imputed) anaylsis (Accessed Feb.4,2017)


```{r}

"%&%" <- function(a,b) paste0(a,b)
library("data.table")
library("dplyr")
library(GenomicRanges)

serv.dir <- "/Users/jtorres/FUSE/"
gwas.dir <- serv.dir %&% "reference/gwas/diagram_1Kgenomes/"
cred.dir <- gwas.dir %&% "credible_sets/"
cred.files <- list.files(cred.dir)[grepl("_window.txt",list.files(cred.dir))]
names(cred.files) <- 1:length(cred.files)

bed.dir <- serv.dir %&% "reference/islet/"

save.dir <- serv.dir %&% "projects/t2d-integration/fGWAS_files/diagram_1Kgenomes/"

diag.df <- fread("cat " %&% gwas.dir %&% "DIAGRAM_T2D_1000G_credSNPs.txt.gz" %&% " |  zmore")
diag.df$MarkerName <- paste0("chr",diag.df$MarkerName)

```


```{r}

build_chunk <- function(refnum){
  fname <- as.character(cred.files[refnum])
  print(fname)
  nam <- gsub("Credset_99_1000G__","",fname)
  nam <- gsub("_201015_LD_window.txt","",nam)
  nam <- gsub("_",":",nam)
  nam <- paste0("chr",nam)
  df <- fread(cred.dir %&% fname)
  df <- select(df, one_of("chrpos_b37","post_prob"))#,"n99"))
  names(df) <- c("SNPID","PPA")#,"SEGNUMBER")
  SEGNUMBER <- rep(refnum,dim(df)[1])
  df <- cbind(df,SEGNUMBER)
  df$SNPID <- paste0("chr",df$SNPID)
  CHR <- as.character(sapply(df$SNPID,function(string){
     tmp <- strsplit(string,split=":")[[1]][1]
     return(tmp)
  }))
  POS <- as.integer(sapply(df$SNPID,function(string){
     tmp <- strsplit(string,split=":")[[1]][2]
     return(tmp)
  }))  
  df <- data.frame(df$SNPID,CHR,POS,df$SEGNUMBER,df$PPA,stringsAsFactors=FALSE)
  names(df) <- c("SNPID","CHR","POS","SEGNUMBER","PPA")
  SE <- as.numeric(sapply(df$SNPID, function(snp){
    temp.df <- filter(diag.df,MarkerName==snp)
    return(temp.df$StdErr)
  }))
  P <- as.numeric(sapply(df$SNPID, function(snp){
    temp.df <- filter(diag.df,MarkerName==snp)
    return(temp.df$`P-value`)
  }))
  Z <- as.numeric(sapply(df$SNPID, function(snp){
    temp.df <- filter(diag.df,MarkerName==snp)
    z <- temp.df$Effect / temp.df$StdErr
    return(z) 
  }))
  df <- cbind(df,Z,SE,P)
  #P <- 2*(1-pnorm(abs(df$Effect/df$StdErr)))  
  df <- arrange(df,POS)
  return(df)
}

build_fgwas_core <- function(){
  out.df <- c() 
  for (i in as.integer(names(cred.files))){
    df <- build_chunk(i)
    out.df <- rbind(out.df,df)
  }
  out.df <- arrange(out.df, SEGNUMBER)
  saveRDS(out.df,file=save.dir%&%"diagram_1KGenomes_fGWAS-core.df.RDS")
  return(out.df)
}

core.df <- build_fgwas_core()

```


Functions for getting annotation columns from annotation files 

```{r}

g <- function(df){
   if(length(df)==3){
      gr <- with(df, GRanges(chr, IRanges(start, end)))
   } else if (length(df)==4){
      gr <- with(df, GRanges(chr, IRanges(start, end), id=id))
   } else if (length(df)==5){
      gr <- with(df, GRanges(chr, IRanges(start, end), id=id, score=score))
   } else if (length(df)>=6){
      df$strand <- gsub(pattern="[^+-]+",replacement='*', df$strand)
      gr <- with(df, GRanges(chr, IRanges(start, end), id=id, score=score, strand=strand))
   }
   return(gr)
}

get_overlaps <- function(a,b){
  # a and b are genomic ranges
  val <- as.integer(a %over% b)
  return(val)
}


get_annot <- function(core.df, annot, prefix, annot.input.df=TRUE){
  pb <- txtProgressBar(min = 0, max = dim(core.df)[1], initial = 0, style = 3)
  vec <- as.integer(sapply(1:dim(core.df)[1], function(i){
    setTxtProgressBar(pb, i) 
    chr <- core.df$CHR[i]
    pos <- core.df$POS[i]
    snp.gr <- GRanges(chr,IRanges(pos, pos))
    if (annot.input.df==TRUE){
      annot.gr <- g(annot)
      value <- get_overlaps(snp.gr,annot.gr)      
    } else{
      value <- get_overlaps(snp.gr,annot)      
    }
    return(value)
  }))
  close(pb)
  saveRDS(vec,file=save.dir%&%prefix%&%".RDS")
  return(vec)
}


```


# Islet ATAC and Chromatin State Annotations 

## Islet ATAC-seq 

Designate BED/annotation files and define function for adding annotation columnns to core.df 

```{r}

prepare_islet_atac <- function(){
  atac.df <- fread(bed.dir %&% "islet_atac_peaks.bed")
  names(atac.df) <- c("chr","start","end","id") 
  atac.df$start <- atac.df$start+1
  atac.df$end <- atac.df$end+1
  return(atac.df)
}

annot_islet_atac <- function(){
  atac.df <- prepare_islet_atac()
  islet_atac <- get_annot(core.df,atac.df, "islet_atac")
  return(islet_atac)
}

#islet_atac <- annot_islet_atac()

```


## Process Islet 15 chromatin state file 

```{r}


prepare_islet_chromHMM_15states <- function(){
  chmm.df <- read.table(bed.dir %&% "Pancreat_islet_15_dense.reformatted_colours.bed", sep="\t",
                      header=FALSE,stringsAsFactors = FALSE)
  chmm.df <- select(chmm.df, one_of("V1","V2","V3","V4"))
  names(chmm.df) <- c("chr","start","end","id") 
  chmm.df$start <- chmm.df$start+1
  chmm.df$end <- chmm.df$end+1
  return(chmm.df)
}

annot_chrom_states <- function(){
  chmm.df <- prepare_islet_chromHMM_15states()
  for (i in 1:15){
    pre <- "islet_state" %&% i
    print(pre)
    sub.df <- filter(chmm.df,id==i)
    annot <- get_annot(core.df,sub.df, pre)    
  }
}

#annot_chrom_states()


```


# TxDb for genomic features 

```{r}

library("GenomicFeatures")
library("TxDb.Hsapiens.UCSC.hg19.knownGene")
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene 
head(seqlevels(txdb))
#seqlevels(txdb) <- "chr1"
#seqlevels(txdb) <- "chr1"
columns(txdb)




```



# Annotation Hub 

```{r}

library(AnnotationHub)
ah = AnnotationHub()
unique(ah$dataprovider)
unique(ah$species)[grepl("Homo",unique(ah$species))] # "Homo sapiens"
dm <- query(ah, c("GRanges", "UCSC", "Homo sapiens"))
df <- mcols(dm)


```



