---
title: "functional_credible_sets.Rmd"
author: "Jason Torres"
date: "May 22, 2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Constructing functional (fgwas) credible sets from the fgwas model with the best cross validated likelihood 

# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)
library("data.table")
library("dplyr")
library("ggplot2")
library(GenomicRanges)

serv.dir <- "/Users/jtorres/FUSE/"

work.dir <- serv.dir %&% "projects/t2d-integration/fgwas/diagram_hrc/"
fgwas.output.dir <- work.dir %&% "fgwas_output/"

cred.set.dir <- serv.dir %&% "projects/t2d-integration/fgwas/diagram_hrc/credible_sets/"

pre <- fgwas.output.dir %&% "drop-islet_state11+islet_state6+islet_state12+utr_5+" %&%
  "islet_stretch_enhancer+promoter+distance_tss+islet_state9"

```


```{r}

get_cred <- function(dframe,cname,prob=0.95){
  index <- match(cname,names(dframe))
  vec <- sort(dframe[,index],decreasing=TRUE)
  count=0
  sum=0
  for (v in vec){
    count <- count + 1
    sum <- sum + v
    if (sum >= prob){
      break
    }
  }
  return(count)
}


```


# Get significant blocks (cummulative PPA > 0.80)

```{r}

get_sig_blocks <- function(ppa.thresh=0.80){
  blk.df <- fread("cat " %&% pre %&% ".segbfs.gz" %&% " | zmore")
  sig.blk.df <- filter(blk.df, PPA >= ppa.thresh) %>% arrange(desc(PPA))
  chrom <- as.integer(gsub("chr","",sig.blk.df$chr))
  sig.blk.df$chrom <- chrom
  sig.blk.df <- arrange(sig.blk.df,chrom,st) %>%
    dplyr::select(one_of("chr","st","sp","PPA"))
  names(sig.blk.df) <- c("CHR","blk.start","blk.end","blk.PPA")
  SEGNUMBER <- 1:dim(sig.blk.df)[1]
  sig.blk.df <- cbind(SEGNUMBER,sig.blk.df)
  write.table(x=sig.blk.df,file = work.dir %&% "sig_blocks_ppa"%&%ppa.thresh%&%".txt", 
              sep="\t",quote=FALSE,row.names=FALSE)
  prop.sig <- dim(sig.blk.df)[1]/dim(blk.df)[1] # 0.0582762
  print("Proportion of blocks with cummulative PPA >= 0.80: " %&% prop.sig)
  return(sig.blk.df)
}

sig.blk.df <- get_sig_blocks()

```


# Subset fgwas SNP output file to SNP in significant blocks 


```{bash}

python subset_to_seg.py pre sig_block_file.gz

```


# Determine functional credible sets (95% and 99% Credible Sets)


```{r}

segsnps.df <- fread("cat " %&% work.dir %&% "sig_block_snps.bfs.txt.gz" %&% " | zmore")
seg.vec <- sort(unique(segsnps.df$SEGNUMBER))

get_credsets <- function(probthresh){
  out.df <- c()
  for (seg in seg.vec){
    temp.df <- filter(segsnps.df,SEGNUMBER==seg) %>% arrange(desc(PPA))
    cumppa <- sum(temp.df$PPA)
    temp.df <- temp.df[1:get_cred(temp.df,"PPA",prob=probthresh*cumppa),]
    temp.df <- dplyr::select(temp.df,-pi,-chunk,-pseudologPO,-pseudoPPA,-V)
    names(temp.df)[c(2,3,4)] <- c("SNPID","CHR","POS")
    temp.df$PPA <- temp.df$PPA/cumppa # rescale ppa to reflect proportion of cummulative sum in block 
    out.df <- rbind(out.df,temp.df)
  }
  return(out.df)
}

cred95.df <- get_credsets(0.95)
cred99.df <- get_credsets(0.99)



```

# Get nearest genes 

```{r}

annot_refGene <- function(cred.df){
  snp.gr <- GRanges(cred.df$CHR,IRanges(cred.df$POS, cred.df$POS))
  
  library(GenomicFeatures)
  hg19.refseq.db <- makeTxDbFromUCSC(genome="hg19", table="refGene")
  refseq.genes<- genes(hg19.refseq.db)
  library(org.Hs.eg.db)
  library(annotate)
  all.geneids <- elementMetadata(refseq.genes)$gene_id
  all.genesymbols <- getSYMBOL(all.geneids, data='org.Hs.eg')
  ref.df <- data.frame(gene.id=all.geneids,symbol=all.genesymbols,
                       stringsAsFactors = FALSE)
  elementMetadata(refseq.genes)$gene_id <- all.genesymbols
  df <- as.data.frame(refseq.genes)
  sub.gr <- GRanges(seqnames=df$seqnames,
                  IRanges(start=df$start,end=df$end),
                  strand=rep("*",dim(df)[1]),
                  name=df$gene_id)
  nearestGenes <- nearest(snp.gr,sub.gr)
  res <- df$gene_id[nearestGenes]
  dist <- distance(snp.gr, refseq.genes[nearestGenes])
  symbol <- res
  cred.df <- cbind(symbol,cred.df)
  
  
  # Sync symbol names for each SEGNUMBER by MOST COMMON gene 
  sync.df <- c()
  for (seg in seg.vec){
    temp.df <- filter(cred.df,SEGNUMBER==seg)
    top <- names(sort(table(temp.df$symbol),decreasing=TRUE))[1]
    temp.df$symbol <- rep(top,dim(temp.df)[1])
    sync.df <- rbind(sync.df,temp.df)
  }
  cred.df <- sync.df  
  return(cred.df)
}
cred95.df <- annot_refGene(cred95.df)
cred99.df <- annot_refGene(cred99.df)


write.table(x=cred95.df,file=cred.set.dir%&%"fgwas_credsets_95_cum80.txt",sep="\t",
            quote=FALSE,row.names=F)

write.table(x=cred99.df,file=cred.set.dir%&%"fgwas_credsets_99_cum80.txt",sep="\t",
            quote=FALSE,row.names=F)

```



# PPA Proportion Annotation Plot functions 


```{r}

get_format_df <- function(cred.df){
  annot.vec <- names(cred.df)[9:length(names(cred.df))]
  
  out.df <- c() 
  for (seg in seg.vec){
    temp.df <- filter(cred.df,SEGNUMBER==seg)
    cumppa <- sum(temp.df$PPA)
    myrow <- c(seg)
    for (annot in annot.vec){
      index <- match(annot,names(temp.df))
      sub.df <- temp.df[,c(1:8,index),]
      annotppa <- sum(sub.df$PPA[as.logical(sub.df[,9])]) / cumppa
      myrow <- c(myrow,annotppa)
    }
    other <- (1-sum(unique(myrow[2:(length(myrow))])))
    other <- ifelse(other>=0,other,0)
    myrow <- c(myrow,other)
    names(myrow) <- c("SEGNUMBER",annot.vec,"other")
    cumsum <- sum((myrow[2:(length(myrow))]))
    myrow[2:length(myrow)] <- myrow[2:length(myrow)] / cumsum # scale to give proper proportion
    out.df <- rbind(out.df,myrow)
  }
  out.df <- as.data.frame(out.df)
  row.names(out.df) <- NULL
  
  format.df <- c() 
  for (i in 1:dim(out.df)[1]){
    myrow <- out.df[i,]
    SEGNUMBER <- rep(myrow$SEGNUMBER,length(2:length(myrow)))
    Annotation <- names(myrow)[2:length(myrow)]
    Proportion <- as.numeric(myrow[2:length(myrow)])
    stack.df <- data.frame(SEGNUMBER,Annotation,Proportion)
    format.df <- rbind(format.df,stack.df)
  }
  
  format.df$Annotation <- factor(format.df$Annotation,                          levels=c("islet_state8","islet_atac","coding",                     "islet_state5","islet_state2","islet_state1","other"))
  
  Symbol <- c()
  for (i in 1:dim(format.df)[1]){
    seg <- format.df$SEGNUMBER[i]
    symb <- filter(cred.df,SEGNUMBER==seg)$symbol[1]
    Symbol <- c(Symbol,symb)
  }
  format.df <- cbind(Symbol,format.df)
  
  isletreg.df <- c()
  for (seg in seg.vec){
    temp.df <- filter(format.df,SEGNUMBER==seg,
                      Annotation=="islet_atac" | Annotation=="islet_state8")
    islet.reg.prop <- sum(temp.df$Proportion)
    gene <- filter(format.df,SEGNUMBER==seg)$Symbol[1]
    stack.df <- data.frame(Symbol=gene,SEGNUMBER=seg,islet.reg.prop)
    isletreg.df <- rbind(isletreg.df,stack.df)
  }
  isletreg.df <- arrange(isletreg.df,(islet.reg.prop))
  PlotID <- paste0(isletreg.df$SEGNUMBER,": ",isletreg.df$Symbol)
  isletreg.df <- cbind(PlotID,isletreg.df)
  
  PlotID <- paste0(format.df$SEGNUMBER,": ",format.df$Symbol)
  format.df <- cbind(PlotID,format.df)
  format.df$PlotID <- factor(format.df$PlotID,levels=isletreg.df$PlotID)
  format.df$SEGNUMBER <- factor(format.df$SEGNUMBER,levels=isletreg.df$SEGNUMBER)
  #format.df$Symbol <- factor(format.df$Symbol,levels=isletreg.df$Symbol)
  
  SNPCOUNT <- c()
  for (i in 1:dim(format.df)[1]){
    seg <- format.df[i,]$SEGNUMBER
    annot <- format.df[i,]$Annotation
    temp.df <- filter(cred.df,SEGNUMBER==seg)
    if (annot !="other"){
      index <- match(annot,names(cred.df))
      #print(c(index,seg))
      snpcount <- sum(temp.df[,index])
    } else{
      snpcount <- dim(temp.df)[1] - 
        sum(as.integer(as.matrix(temp.df[,8:dim(temp.df)[2]])))
      if (snpcount < 0){
        snpcount <- 0 
      }
    }
    SNPCOUNT <- c(SNPCOUNT,snpcount)
  }
  format.df$SNPCOUNT <- SNPCOUNT 

  return(format.df)
}

ppa_plot <- function(format.df){
  plt <- ggplot(data=format.df,aes(x=PlotID,y=Proportion,fill=Annotation)) + 
            geom_bar(stat="identity",color="black",size=0.2) + 
            scale_y_continuous(breaks=seq(0,1,0.1)) + 
            xlab("Locus") + 
            coord_flip() + 
            scale_fill_brewer(palette="Spectral") +
            ggtitle("Proportion of PPA attributable to each annotation")
  return(plt)
}

snpcount_plot <- function(format.df){
  plt <- ggplot(data=format.df,aes(x=PlotID,y=SNPCOUNT,fill=Annotation)) + 
            geom_bar(stat="identity",color="black",size=0.2) + 
            scale_y_continuous(breaks=seq(0,300,50)) + 
            xlab("Locus")+
            coord_flip() + 
            scale_fill_brewer(palette="Spectral") +
            ggtitle("Proportion of SNP count attributable to each annotation")  
  return(plt)
}


```


95% Credible Sets  

```{r}

form95 <- get_format_df(cred95.df)
ppa.plt.95 <- ppa_plot(form95)
sc.plt.95 <- snpcount_plot(form95)

```


99% Credible Sets 

```{r}

form99 <- get_format_df(cred99.df)
ppa.plt.99 <- ppa_plot(form99)
sc.plt.99 <- snpcount_plot(form99)

```

